mm2src/coins/nft/storage/sql_storage.rs

```diff
@@ -0,0 +1,889 @@
+    );
+    Ok(sql)
+}
+
+fn get_nft_metadata_sql(chain: &Chain) -> MmResult<String, SqlError> {
+    let table_name = nft_list_table_name(chain);
+    validate_table_name(&table_name)?;
+    let sql = format!(
+        "SELECT details_json FROM {} WHERE token_address=?1 AND token_id=?2",
+        table_name
+    );
+    Ok(sql)
+}
+
+fn select_last_block_number_sql<F>(chain: &Chain, table_name_creator: F) -> MmResult<String, SqlError>
+where
+    F: FnOnce(&Chain) -> String,
+{
+    let table_name = table_name_creator(chain);
+    validate_table_name(&table_name)?;
+    let sql = format!(
+        "SELECT block_number FROM {} ORDER BY block_number DESC LIMIT 1",
+        table_name
+    );
+    Ok(sql)
+}
+
+fn select_last_scanned_block_sql() -> MmResult<String, SqlError> {
+    let table_name = scanned_nft_blocks_table_name();
+    validate_table_name(&table_name)?;
+    let sql = format!("SELECT last_scanned_block FROM {} WHERE chain=?1", table_name,);
+    Ok(sql)
+}
+
+fn get_nft_amount_sql<F>(chain: &Chain, table_name_creator: F) -> MmResult<String, SqlError>
+where
+    F: FnOnce(&Chain) -> String,
+{
+    let table_name = table_name_creator(chain);
+    validate_table_name(&table_name)?;
+    let sql = format!(
+        "SELECT amount FROM {} WHERE token_address=?1 AND token_id=?2",
+        table_name
+    );
+    Ok(sql)
+}
+
+fn delete_nft_sql<F>(chain: &Chain, table_name_creator: F) -> Result<String, MmError<SqlError>>
+where
+    F: FnOnce(&Chain) -> String,
+{
+    let table_name = table_name_creator(chain);
+    validate_table_name(&table_name)?;
+    let sql = format!("DELETE FROM {} WHERE token_address=?1 AND token_id=?2", table_name);
+    Ok(sql)
+}
+
+fn block_number_from_row(row: &Row<'_>) -> Result<i64, SqlError> { row.get::<_, i64>(0) }
+
+fn nft_amount_from_row(row: &Row<'_>) -> Result<String, SqlError> { row.get(0) }
+
+fn get_txs_from_block_builder<'a>(
+    conn: &'a Connection,
+    chain: &'a Chain,
+    from_block: u64,
+) -> MmResult<SqlQuery<'a>, SqlError> {
+    let table_name = nft_tx_history_table_name(chain);
+    validate_table_name(table_name.as_str())?;
+    let mut sql_builder = SqlQuery::select_from(conn, table_name.as_str())?;
+    sql_builder
+        .sql_builder()
+        .and_where(format!("block_number >= '{}'", from_block))
+        .order_asc("block_number")
+        .field("details_json");
+    drop_mutability!(sql_builder);
+    Ok(sql_builder)
+}
+
+fn get_txs_by_token_addr_id_builder<'a>(
+    conn: &'a Connection,
+    chain: &'a Chain,
+    token_address: String,
+    token_id: String,
+) -> MmResult<SqlQuery<'a>, SqlError> {
+    let table_name = nft_tx_history_table_name(chain);
+    validate_table_name(table_name.as_str())?;
+    let mut sql_builder = SqlQuery::select_from(conn, table_name.as_str())?;
+    sql_builder
+        .sql_builder()
+        .and_where_eq("token_address", format!("'{}'", token_address))
+        .and_where_eq("token_id", format!("'{}'", token_id))
```

Comment:
This function has potential for sql injection if the moralis endpoint were to somehow return malicious `token_address` values or if it were ever reused with a different source for the `token_id` argument. 

eg,
`let token_id = "a_token_id'; DROP TABLE an_important_table;'".to_string();`
or
``let token_address = "a_token_address'; DROP TABLE an_important_table;'".to_string();``

We should be deserializing any data from external sources to strict types. For example, if we know `token_address` must be an EVM compatible address, we should deserialize it directly to a `Address` type and convert to String as needed to ensure we are in fact dealing with a valid address.

Reply:
Please use prepared statement here @laruh 

We should avoid using `sql_builder` as it will be replaced or refactored to be built on prepared statements. Any query created with `sql_builder` is vulnerable to SQL injections.

Reply:
Ok, on it.

Reply:
Done [here](https://github.com/KomodoPlatform/komodo-defi-framework/pull/1914)